src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -0,0 +1,1312 @@
+        chip::GroupId prev_group = kUndefinedGroupId;
+        GroupInfo group_info;
+
+        // Loop through the endpoint's groups
+        do
+        {
+            // Load next group
+            VerifyOrReturnError(CHIP_NO_ERROR == GroupMapStore(fabric_index, endpoint_id, group_id).Load(group_info), false);
+            if (group_id == mapping.group)
+            {
+                // Target group found
+                return true;
+            }
+            prev_group = group_id;
+            group_id   = group_info.next;
+        } while ((kUndefinedGroupId != group_id) && (prev_group != group_id));
+
+        // Not found
+        return false;
+    }
+
+    CHIP_ERROR AddGroupMapping(chip::FabricIndex fabric_index, const GroupMapping & mapping) override
+    {
+        VerifyOrReturnError(mInitialized, CHIP_ERROR_INTERNAL);
+        VerifyOrReturnError(kInvalidEndpointId != mapping.endpoint, CHIP_ERROR_INVALID_ARGUMENT);
+        VerifyOrReturnError(IsFabricGroupId(mapping.group), CHIP_ERROR_INVALID_ARGUMENT);
+
+        FabricStore fabric_store(fabric_index);
+        FabricInfo fabric_info;
+
+        CHIP_ERROR err = fabric_store.Load(fabric_info);
+        if (CHIP_NO_ERROR != err || kInvalidEndpointId == fabric_info.first_endpoint)
+        {
+            // New fabric, or new endpoint
+            GroupMapStore(fabric_index, mapping.endpoint, mapping.group).Save(GroupInfo(mapping.name));
+            EndpointStore(fabric_index, mapping.endpoint).Save(EndpointInfo(mapping.group));
+            fabric_info.first_endpoint = mapping.endpoint;
+            fabric_info.endpoint_count = 1;
+            return fabric_store.Save(fabric_info);
+        }
+
+        // Existing fabric
+
+        chip::EndpointId endpoint_id   = fabric_info.first_endpoint;
+        chip::EndpointId prev_endpoint = kInvalidEndpointId;
+        EndpointInfo endpoint_info;
+        EndpointInfo prev_end_info;
+        size_t endpoint_count = 0;
+
+        // Loop through the fabric's endpoints
+        do
+        {
+            EndpointStore endpoint_store(fabric_index, endpoint_id);
+            if (CHIP_NO_ERROR != endpoint_store.Load(endpoint_info))
+            {
+                // Endpoint info not found
+                break;
+            }
+            if (endpoint_id == mapping.endpoint)
+            {
+                // Target endpoint info found
+                break;
+            }
+            prev_end_info = endpoint_info;
+            prev_endpoint = endpoint_id;
+            endpoint_id   = endpoint_info.next;
+        } while (++endpoint_count < fabric_info.endpoint_count);
+
+        if (endpoint_id != mapping.endpoint)
+        {
+            // Endpoint info NOT found, create new
+            if (kInvalidEndpointId != prev_endpoint)
+            {
+                // Link to existing endpoint
+                prev_end_info.next = mapping.endpoint;
+                EndpointStore(fabric_index, prev_endpoint).Save(prev_end_info);
```

Comment:
If we inserted at the start of the list, not the end, we'd instead need to store the `FabricInfo` here, but we'd avoid having to track `prev_end_info` and `prev_endpoint` in the loop above, right?  might be a bit simpler...

Reply:
When you're right, you're right. I am inserting now the new elements first, and updated the unit tests to not expect any particular order.